targets/TARGET_RDA/TARGET_UNO_91H/trng_api.c

```diff
@@ -0,0 +1,111 @@
+/*
+ *  Hardware entropy collector for the UNO_91H, using RDA's TRNG
+ *
+ *  Copyright (C) 2006-2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#if defined(DEVICE_TRNG)
+
+#include "cmsis.h"
+#include "trng_api.h"
+#include "us_ticker_api.h"
+
+#define rRNG_CLKGATE        (RDA_SCU->CLKGATE0)
+#define rRNG_CLKGEN         (RDA_DMACFG->dma_func_ctrl)
+#define rRNG_INTSTATE       (RDA_DMACFG->dma_int_out)
+#define rTRNG_CTRL          (RDA_RNG->TCTRL)
+#define rTRNG_DATA0         (RDA_RNG->TD0)
+#define rTRNG_DATA1         (RDA_RNG->TD1)
+#define rPRNG_DATA          (RDA_RNG->PD)
+
+static uint32_t trng_inited = 0;
+
+/** trng_get_byte
+ *  @brief Get one byte of entropy from the RNG, assuming it is up and running.
+ *  @param obj TRNG obj
+ *  @param pointer to the hardware generated random byte.
+ */
+static void trng_get_byte(trng_t *obj, unsigned char *byte )
+{
+    if (0U == obj->byte_idx) {
+        uint32_t start_time = us_ticker_read();
+        /* Delay for PRNG data ready, generate 4-byte per 32us */
+        while ((us_ticker_read() - start_time) <= 32);
+    }
+    *byte = (unsigned char)((rPRNG_DATA >> (obj->byte_idx << 3)) & 0x00FFUL);
```

Comment:
If we can't get an entropy estimate for the TRNG output, then I suggest we work with the assumption that every 32 bit value in the register has a single bit of entropy. (Of course we can't exclude the possibility that this is still to optimistic, but I would consider it relatively safe)

The Mbed HAL TRNG API requires to pass full entropy in the buffer so we need to xor all the bits of the register together to gain 1 bit of entropy that we can pass in the output.

Reply:
@yanesca I have updated it in the lastest PR, you can found it in trng_init(). Thanks.

Reply:
@kyliuxing I am sorry but I can't find the comment in `trng_init()`, can you give me a line number?